contracts/dual_token/DualStakingToken.sol

```diff
@@ -0,0 +1,362 @@
+            uint256 totalRewards;
+            for (uint256 i = 0; i < numServices; ++i) {
+                totalRewards += eligibleServiceRewards[i];
+            }
+
+            uint256 curServiceId;
+            // If total allocated rewards are not enough, adjust the reward value
+            if ((totalRewards * rewardRatio) / 1e18 > lastAvailableRewards) {
+                // Traverse all the eligible services and adjust their rewards proportional to leftovers
+                // Note the algorithm is the exact copy of StakingBase logic
+                uint256 updatedReward;
+                uint256 updatedTotalRewards;
+                for (uint256 i = 1; i < numServices; ++i) {
+                    // Calculate the updated reward
+                    updatedReward = (eligibleServiceRewards[i] * lastAvailableRewards) / totalRewards;
+                    // Add to the total updated reward
+                    updatedTotalRewards += updatedReward;
+
+                    curServiceId = eligibleServiceIds[i];
+                    // Add reward to the overall service reward
+                    mapStakerStakerInfos[curServiceId].reward += (updatedReward * rewardRatio) / 1e18;
+                }
+
+                // Process the first service in the set
+                updatedReward = (eligibleServiceRewards[0] * lastAvailableRewards) / totalRewards;
+                updatedTotalRewards += updatedReward;
+                curServiceId = eligibleServiceIds[0];
+                // If the reward adjustment happened to have small leftovers, add it to the first service
+                if (lastAvailableRewards > updatedTotalRewards) {
+                    updatedReward += lastAvailableRewards - updatedTotalRewards;
+                }
+                // Add reward to the overall service reward
+                mapStakerStakerInfos[curServiceId].reward += (updatedReward * rewardRatio) / 1e18;
+                // Set available rewards to zero
+                lastAvailableRewards = 0;
+            } else {
+                // Traverse all the eligible services and add to their rewards
+                for (uint256 i = 0; i < numServices; ++i) {
+                    // Add reward to the service overall reward
+                    curServiceId = eligibleServiceIds[i];
+                    mapStakerStakerInfos[curServiceId].reward += (eligibleServiceRewards[i] * rewardRatio) / 1e18;
+                }
+
+                // Adjust available rewards
+                lastAvailableRewards -= totalRewards;
+            }
+
+            // Update the storage value of available rewards
+            availableRewards = lastAvailableRewards;
+        }
+
+        // Lock activity checker after checkpoint
+        IActivityChecker(activityChecker).lock();
+    }
+
+    /// @dev Re-stakes OLAS service Id as it has been evicted.
+    /// @param serviceId OLAS driven service Id.
+    function restake(uint256 serviceId) external {
```

Comment:
restake only from staking instance is prevented by the verifier via the checkpoint, correct?

Reply:
Yes, the restake can only take place is the service is evicted, otherwise it will revert. So if the service is already unstaked from the original staking instance, this function will revert and the only way to sync the execution is to call the final unstake, then stake from scratch.

Reply:
Please add to docstring